backend/services/operators.py

```diff
@@ -0,0 +1,191 @@
+        if len(gene1) != len(gene2):
+            raise ValueError("Chromosomy muszą mieć taką samą długość.")
+        point = random.randint(1, len(gene1) - 1)
+        child_gene1 = gene1[:point] + gene2[point:]
+        child_gene2 = gene2[:point] + gene1[point:]
+        return Individual(Chromosome(child_gene1)), Individual(Chromosome(child_gene2))
+
+class TwoPointCrossover(CrossoverOperator):
+    def crossover(self, parent1: Individual, parent2: Individual):
+        gene1 = parent1.chromosome.gene
+        gene2 = parent2.chromosome.gene
+        if len(gene1) != len(gene2):
+            raise ValueError("Chromosomy muszą mieć taką samą długość.")
+        if len(gene1) < 2:
+            return OnePointCrossover().crossover(parent1, parent2)
+        point1, point2 = sorted(random.sample(range(1, len(gene1)), 2))
+        child_gene1 = gene1[:point1] + gene2[point1:point2] + gene1[point2:]
+        child_gene2 = gene2[:point1] + gene1[point1:point2] + gene2[point2:]
+        return Individual(Chromosome(child_gene1)), Individual(Chromosome(child_gene2))
+
+class UniformCrossover(CrossoverOperator):
+    def crossover(self, parent1: Individual, parent2: Individual):
+        gene1 = parent1.chromosome.gene
+        gene2 = parent2.chromosome.gene
+        if len(gene1) != len(gene2):
+            raise ValueError("Chromosomy muszą mieć taką samą długość.")
+        child_gene1 = ''.join(g1 if random.random() < 0.5 else g2 for g1, g2 in zip(gene1, gene2))
+        child_gene2 = ''.join(g2 if random.random() < 0.5 else g1 for g1, g2 in zip(gene1, gene2))
+        return Individual(Chromosome(child_gene1)), Individual(Chromosome(child_gene2))
+
+class GrainCrossover(CrossoverOperator):
+    def crossover(self, parent1: Individual, parent2: Individual):
+        gene1 = parent1.chromosome.gene
+        gene2 = parent2.chromosome.gene
+        if len(gene1) != len(gene2):
+            raise ValueError("Chromosomy muszą mieć taką samą długość.")
+        
+        child_gene1 = ""
+        child_gene2 = ""
+        for i in range(len(gene1)):
+            # Dla potomka 1 losujemy osobno dla każdego genu
+            if random.random() <= 0.5:
+                child_gene1 += gene1[i]
+            else:
+                child_gene1 += gene2[i]
+            # Dla potomka 2 również losujemy niezależnie
+            if random.random() <= 0.5:
+                child_gene2 += gene1[i]
+            else:
+                child_gene2 += gene2[i]
+                
+        return Individual(Chromosome(child_gene1)), Individual(Chromosome(child_gene2))
+
+
+# --- MUTACJA ---
+
+class MutationOperator(ABC):
+    @abstractmethod
+    def mutate(self, individual: Individual, mutation_probability: float):
+        """
+        Mutuje danego osobnika z podanym prawdopodobieństwem.
+        """
+        pass
+
+class OnePointMutation(MutationOperator):
+    def mutate(self, individual: Individual, mutation_probability: float):
```

Comment:
Jak dla mnie torchę dziwne jest, że poprzednie operatory nie uwzgledniaja prawdopodobieństwa, a te tak. Ale spoko, jak już tak jest to niech tak będzie